### PR TITLE
JBIDE-25380

### PR DESCRIPTION
**Pull request by robstryker**
# Pull Request Checklist
## General
* Is this a blocking issue or new feature? If yes, QE needs to +1 this PR

## Code
* Are method-/class-/variable-names meaningful?
* Are methods concise, not too long?
* Are catch blocks catching precise Exceptions only (no catch all)?

## Testing
* Are there unit-tests?
* Are there integration tests (or at least a jira to tackle these)?
* Is the non-happy path working, too?
* Are other parts that use the same component still working fine?

## Function
* Does it work?

### CODE REVIEW

**Comment by robstryker** [comment]
testPR

File: plugins/org.jboss.tools.openshift.ui/src/org/jboss/tools/openshift/internal/ui/server/OpenShiftServerEditorSection.java [hunk 1]
```diff
@@ -274,176 +195,43 @@ protected void createEclipseProjectControls(FormToolkit toolkit, Composite paren
...
-				})
-				.to(BeanProperties.value(OpenShiftServerEditorModel.PROPERTY_SOURCE_PATH).observe(model))
-				.in(dbc);
-
+		applyAlignData(SWT.BEGINNING, SWT.CENTER,sourcePathLabel);
+
+		sourcePathText = new Text(parent, SWT.BORDER | SWT.READ_ONLY);
+		applyGridData(2, 1, SWT.FILL, SWT.FILL,true, false, sourcePathText);
+		// TODO validate. User can't change this value though, so only needs validation once
+//		if (!(value instanceof String) || StringUtils.isEmpty(value)) {
```
Comment (adietish) [inline plugins/org.jboss.tools.openshift.ui/src/org/jboss/tools/openshift/internal/ui/server/OpenShiftServerEditorSection.java]: still TODO?

File: plugins/org.jboss.tools.openshift.ui/src/org/jboss/tools/openshift/internal/ui/server/OpenShiftServerEditorSection.java [hunk 2]
```diff
@@ -452,302 +240,403 @@ private void createOpenShiftApplicationControls(FormToolkit toolkit, Composite p
...
-		Button selectResourceButton = new Button(parent, SWT.PUSH);
-		selectResourceButton.setText("Select...");
-		GridDataFactory.fillDefaults()
-				.align(SWT.FILL, SWT.CENTER).hint(120, SWT.DEFAULT)
-				.applyTo(selectResourceButton);
-		selectResourceButton.addSelectionListener(onSelectResource(model, selectResourceButton.getShell()));
+		resourceKindLabel = new Label(parent, SWT.None);
+		applyAlignData(SWT.FILL, SWT.CENTER, resourceKindLabel);
+		resourceText = new Text(parent, SWT.BORDER | SWT.READ_ONLY);
+		applyGridData(2, 1, SWT.FILL, SWT.CENTER, true, false, resourceText);
```
Comment (adietish) [inline plugins/org.jboss.tools.openshift.ui/src/org/jboss/tools/openshift/internal/ui/server/OpenShiftServerEditorSection.java]: that's really ugly. There's no way for you to know what these parameters are about without getting to the method and reading what it really does.

File: plugins/org.jboss.tools.openshift.ui/src/org/jboss/tools/openshift/internal/ui/server/OpenShiftServerEditorSection.java [hunk 3]
```diff
@@ -452,302 +240,403 @@ private void createOpenShiftApplicationControls(FormToolkit toolkit, Composite p
...
+			@Override
+			public IStatus validate(Object value) {
+				if( !useImageDevmodeKeyButton.getSelection()) {
+					return new OpenShiftIdentifierValidator().validate(devmodeKeyText.getText());
+				}
+				return Status.OK_STATUS;
+			}
+		});
+		devmodeKeyText.addModifyListener(devmodeKeyModifyDecoration);
+
```
Comment (adietish) [inline plugins/org.jboss.tools.openshift.ui/src/org/jboss/tools/openshift/internal/ui/server/OpenShiftServerEditorSection.java]: 2 empty lines?

File: plugins/org.jboss.tools.openshift.ui/src/org/jboss/tools/openshift/internal/ui/server/OpenShiftServerEditorSection.java [hunk 4]
```diff
@@ -452,302 +240,403 @@ private void createOpenShiftApplicationControls(FormToolkit toolkit, Composite p
...
+		
+		
+		useImageDevmodeKeyListener = new SelectionAdapter() {
+			@Override
+			public void widgetSelected(SelectionEvent e) {
+				execute(new ToggleDebuggingCommand(input.getServer(), useImageDevmodeKeyButton, devmodeKeyText, this, devmodeKeyModifyListener));
+			}
+		};
+	}
+
```
Comment (adietish) [inline plugins/org.jboss.tools.openshift.ui/src/org/jboss/tools/openshift/internal/ui/server/OpenShiftServerEditorSection.java]: 2 empty lines?

File: plugins/org.jboss.tools.openshift.ui/src/org/jboss/tools/openshift/internal/ui/server/OpenShiftServerEditorSection.java [hunk 5]
```diff
@@ -452,302 +240,403 @@ private void createOpenShiftApplicationControls(FormToolkit toolkit, Composite p
...
+		String preText;
+		
+		public ToggleDebuggingCommand(IServerWorkingCopy server, Button useImageDevmodeKeyButton, Text devmodeKeyText,
+				SelectionListener toggleListener, ModifyListener textListener) {
+			super(server, "Enable Debugging");
+			this.wc = server;
+			this.useImageDevmodeKeyButton = useImageDevmodeKeyButton;
+			this.devmodeKeyText = devmodeKeyText;
+			preSelected = !useImageDevmodeKeyButton.getSelection();
+			preText = devmodeKeyText.getText();
```
Comment (adietish) [inline plugins/org.jboss.tools.openshift.ui/src/org/jboss/tools/openshift/internal/ui/server/OpenShiftServerEditorSection.java]: preText? not the most speaking variable name imho

File: plugins/org.jboss.tools.openshift.ui/src/org/jboss/tools/openshift/internal/ui/server/OpenShiftServerEditorSection.java [hunk 6]
```diff
@@ -452,302 +240,403 @@ private void createOpenShiftApplicationControls(FormToolkit toolkit, Composite p
...
+	
+
+	private static class ToggleDebuggingCommand extends ServerCommand {
+		private IServerWorkingCopy wc;
+		private Button useImageDevmodeKeyButton;
+		private Text devmodeKeyText;
+		private SelectionListener toggleListener;
+		private ModifyListener textListener;
+		boolean preSelected;
+		String preText;
```
Comment (adietish) [inline plugins/org.jboss.tools.openshift.ui/src/org/jboss/tools/openshift/internal/ui/server/OpenShiftServerEditorSection.java]: why these 2 default accessibility while the others are private?

File: plugins/org.jboss.tools.openshift.ui/src/org/jboss/tools/openshift/internal/ui/server/OpenShiftServerEditorSection.java (hunk 5 above)
Comment (adietish) [reply plugins/org.jboss.tools.openshift.ui/src/org/jboss/tools/openshift/internal/ui/server/OpenShiftServerEditorSection.java]: why this mix of using this. and not?

File: plugins/org.jboss.tools.openshift.ui/src/org/jboss/tools/openshift/internal/ui/server/OpenShiftServerEditorSection.java [hunk 7]
```diff
@@ -452,302 +240,403 @@ private void createOpenShiftApplicationControls(FormToolkit toolkit, Composite p
...
+		}; 
+		ModifyListener deployPathDecoration = new TextboxDecoratorSupport(deployPathText, new IValidator() {
+			@Override
+			public IStatus validate(Object value) {
+				// TODO validate 
+				if (!(value instanceof String) || StringUtils.isEmpty(value)) {
+					return ValidationStatus.cancel("Please provide a path to deploy to on the pod.");
+				}
+				if (!Path.isValidPosixPath((String)value)) {
+					return ValidationStatus.error("Please provide a valid path to deploy to on the pod");
```
Comment (adietish) [inline plugins/org.jboss.tools.openshift.ui/src/org/jboss/tools/openshift/internal/ui/server/OpenShiftServerEditorSection.java]: can be written in a more jdk8 like fashion with lambda:
```
		ModifyListener deployPathDecoration = new TextboxDecoratorSupport(deployPathText, 
			value -> {
				// TODO validate 
				if (!(value instanceof String) || StringUtils.isEmpty(value)) {
					return ValidationStatus.cancel("Please provide a path to deploy to on the pod.");
				}
				if (!Path.isValidPosixPath((String)value)) {
					return ValidationStatus.error("Please provide a valid path to deploy to on the pod");
				}
				return Status.OK_STATUS;
		});
```
like many of those listeners that you use.

File: plugins/org.jboss.tools.openshift.ui/src/org/jboss/tools/openshift/internal/ui/server/OpenShiftServerEditorSection.java [hunk 8]
```diff
@@ -452,302 +240,403 @@ private void createOpenShiftApplicationControls(FormToolkit toolkit, Composite p
...
+		
+		devmodeKeyModifyListener = new ModifyListener() {
+			public void modifyText(ModifyEvent e) {
+				execute(new ServerWorkingCopyPropertyTextCommand(input.getServer(), "SetDevmodeKey", devmodeKeyText, 
+						devmodeKeyText.getText(), OpenShiftServerUtils.ATTR_DEVMODE_KEY, this));
+			}
+		}; 
+		
+
+		ModifyListener devmodeKeyModifyDecoration = new TextboxDecoratorSupport(devmodeKeyText, new IValidator() {
```
Comment (adietish) [inline plugins/org.jboss.tools.openshift.ui/src/org/jboss/tools/openshift/internal/ui/server/OpenShiftServerEditorSection.java]: lambda?

File: plugins/org.jboss.tools.openshift.ui/src/org/jboss/tools/openshift/internal/ui/server/OpenShiftServerEditorSection.java [hunk 9]
```diff
@@ -452,302 +240,403 @@ private void createOpenShiftApplicationControls(FormToolkit toolkit, Composite p
...
+
+		ModifyListener debugPortKeyDecoration = new TextboxDecoratorSupport(debugPortKeyText, new IValidator() {
+			@Override
+			public IStatus validate(Object value) {
+				if( !useImageDebugPortKeyButton.getSelection()) {
+					return new OpenShiftIdentifierValidator().validate(debugPortKeyText.getText());
+				}
+				return Status.OK_STATUS;
+			}
+		});
```
Comment (adietish) [inline plugins/org.jboss.tools.openshift.ui/src/org/jboss/tools/openshift/internal/ui/server/OpenShiftServerEditorSection.java]: lambda?

File: plugins/org.jboss.tools.openshift.ui/src/org/jboss/tools/openshift/internal/ui/server/OpenShiftServerEditorSection.java [hunk 10]
```diff
@@ -452,302 +240,403 @@ private void createOpenShiftApplicationControls(FormToolkit toolkit, Composite p
...
-				if (WizardUtils.openWizardDialog(selectResourceDialog, shell) 
-						== Dialog.OK) {
-					model.setResource(selectResourceDialog.getResource());
+			public IStatus validate(Object value) {
+				if( !useImageDebugPortKeyButton.getSelection()) {
+					return new NumericValidator("integer", Integer::parseInt, true).validate(debugPortText.getText());
 				}
+				return Status.OK_STATUS;
+			}
+		});
```
Comment (adietish) [inline plugins/org.jboss.tools.openshift.ui/src/org/jboss/tools/openshift/internal/ui/server/OpenShiftServerEditorSection.java]: lambda?

File: plugins/org.jboss.tools.openshift.ui/src/org/jboss/tools/openshift/internal/ui/server/OpenShiftServerEditorSection.java [hunk 11]
```diff
@@ -452,302 +240,403 @@ private void createOpenShiftApplicationControls(FormToolkit toolkit, Composite p
...
+		debugPortKeyValSelectionListener = new SelectionAdapter() {
+
+			@Override
+			public void widgetSelected(SelectionEvent e) {
+				execute(new ToggleDebugKeyValueCommand(server, useImageDebugPortKeyButton, 
+						this, debugPortKeyText, debugPortKeyListener, debugPortText, debugPortValListener));
 			}
 		};
+		
+//		ControlDecorationSupport.create(debugPortBinding,SWT.LEFT | SWT.TOP, container
```
Comment (adietish) [inline plugins/org.jboss.tools.openshift.ui/src/org/jboss/tools/openshift/internal/ui/server/OpenShiftServerEditorSection.java]: What's with these commented lines?

File: plugins/org.jboss.tools.openshift.ui/src/org/jboss/tools/openshift/internal/ui/server/OpenShiftServerEditorSection.java [hunk 12]
```diff
@@ -452,302 +240,403 @@ private void createOpenShiftApplicationControls(FormToolkit toolkit, Composite p
...
+		            decorator.setImage(image);
+		            decorator.setDescriptionText(s.getMessage());
+		            decorator.show();
+	            } else {
+	            	decorator.hide();
+	            }
+			} else {
+				decorator.hide();
+			}
+		}
```
Comment (adietish) [inline plugins/org.jboss.tools.openshift.ui/src/org/jboss/tools/openshift/internal/ui/server/OpenShiftServerEditorSection.java]: can you please correct the weird identations here? it's very hard to read.

File: plugins/org.jboss.tools.openshift.ui/src/org/jboss/tools/openshift/internal/ui/server/OpenShiftServerEditorSection.java [hunk 13]
```diff
@@ -452,302 +240,403 @@ private void createOpenShiftApplicationControls(FormToolkit toolkit, Composite p
...
+	private static class ToggleDebugKeyValueCommand extends ServerCommand {
+		private IServerWorkingCopy wc;
+		private Button useImageDebugPortKeyButton;
+		private SelectionListener toggleListener;
+		
+		private Text debugPortKeyText, debugPortText;
+		private ModifyListener keyModifier, valModifier;
+		
+		
+		boolean preSelected;
```
Comment (adietish) [inline plugins/org.jboss.tools.openshift.ui/src/org/jboss/tools/openshift/internal/ui/server/OpenShiftServerEditorSection.java]: default? why not private?

File: plugins/org.jboss.tools.openshift.ui/src/org/jboss/tools/openshift/internal/ui/server/OpenShiftServerEditorSection.java [hunk 14]
```diff
@@ -452,302 +240,403 @@ private void createOpenShiftApplicationControls(FormToolkit toolkit, Composite p
...
+			this.toggleListener = toggleListener;
+			
+			this.debugPortKeyText = debugPortKeyText;
+			this.debugPortText = debugPortText;
+			this.keyModifier = keyModifier;
+			this.valModifier = valModifier;
+			
+			preSelected = !useImageDebugPortKeyButton.getSelection();
+			preKey = debugPortKeyText.getText();
+			preVal = debugPortText.getText();
```
Comment (adietish) [inline plugins/org.jboss.tools.openshift.ui/src/org/jboss/tools/openshift/internal/ui/server/OpenShiftServerEditorSection.java]: why mixing using this/no this?

Comment (adietish) [reply plugins/org.jboss.tools.openshift.ui/src/org/jboss/tools/openshift/internal/ui/server/OpenShiftServerEditorSection.java]: is there really no more speaking var name than preKey, preVal?

**Comment by robstryker** [comment]
cleaned it all up. Going for 2nd approval

**Comment by adietish** [comment]
if you have a server adapter that points to a cdk that you dont have any more (different IP), the you get the following NPE:
image
```
java.lang.NullPointerException
	at org.jboss.tools.openshift.internal.ui.server.OpenShiftServerEditorSection$InitializeWidgetsJob.initializeWidgetValues(OpenShiftServerEditorSection.java:621)
	at org.jboss.tools.openshift.internal.ui.server.OpenShiftServerEditorSection$InitializeWidgetsJob.runInUIThread(OpenShiftServerEditorSection.java:604)
	at org.eclipse.ui.progress.UIJob.lambda$0(UIJob.java:95)
	at org.eclipse.swt.widgets.RunnableLock.run(RunnableLock.java:37)
	at org.eclipse.swt.widgets.Synchronizer.runAsyncMessages(Synchronizer.java:182)
	at org.eclipse.swt.widgets.Display.runAsyncMessages(Display.java:4033)
	at org.eclipse.swt.widgets.Display.readAndDispatch(Display.java:3700)
	at org.eclipse.e4.ui.internal.workbench.swt.PartRenderingEngine$5.run(PartRenderingEngine.java:1150)
	at org.eclipse.core.databinding.observable.Realm.runWithDefault(Realm.java:336)
	at org.eclipse.e4.ui.internal.workbench.swt.PartRenderingEngine.run(PartRenderingEngine.java:1039)
	at org.eclipse.e4.ui.internal.workbench.E4Workbench.createAndRunUI(E4Workbench.java:153)
	at org.eclipse.ui.internal.Workbench.lambda$3(Workbench.java:680)
	at org.eclipse.core.databinding.observable.Realm.runWithDefault(Realm.java:336)
	at org.eclipse.ui.internal.Workbench.createAndRunWorkbench(Workbench.java:594)
	at org.eclipse.ui.PlatformUI.createAndRunWorkbench(PlatformUI.java:148)
	at org.eclipse.ui.internal.ide.application.IDEApplication.start(IDEApplication.java:151)
	at org.eclipse.equinox.internal.app.EclipseAppHandle.run(EclipseAppHandle.java:196)
	at org.eclipse.core.runtime.internal.adaptor.EclipseAppLauncher.runApplication(EclipseAppLauncher.java:134)
	at org.eclipse.core.runtime.internal.adaptor.EclipseAppLauncher.start(EclipseAppLauncher.java:104)
	at org.eclipse.core.runtime.adaptor.EclipseStarter.run(EclipseStarter.java:388)
	at org.eclipse.core.runtime.adaptor.EclipseStarter.run(EclipseStarter.java:243)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.eclipse.equinox.launcher.Main.invokeFramework(Main.java:653)
	at org.eclipse.equinox.launcher.Main.basicRun(Main.java:590)
	at org.eclipse.equinox.launcher.Main.run(Main.java:1499)
	at org.eclipse.equinox.launcher.Main.main(Main.java:1472)
```

**Pull request merged**